Add M2M Oauth to Databricks Driver

AlexVialaBellander opened the pull request:
# Add M2M Oauth to Databricks Driver
Closes https://github.com/metabase/metabase/issues/51276

### Description
This pull request includes updates to the Databricks drivers and its documentation. The changes enhance the connection setup process and provide more detailed guidance on authentication methods and configuration options.

#### Documentation updates:

* [`docs/databases/connections/databricks.md`](diffhunk://#diff-0498b9e0ba7072177771a34a8ebdfddaf0a0af20d4cb4a6dc95eed9540ff6b3dL17-R39): Expanded the host and HTTP path descriptions with examples and added detailed sections on authentication methods, re-running queries, and managing database sync and scan settings. 

#### Driver configuration updates:

* `modules/drivers/databricks/resources/metabase-plugin.yaml`: Updated the driver version to 1.2.0, added placeholders and helper text for various connection properties, and included new options for OAuth M2M authentication.
* `modules/drivers/databricks/src/metabase/driver/databricks.clj`: Modified the connection details to support both personal access tokens and OAuth M2M authentication, with conditional logic to toggle between OAuth M2M and token based auth. M2M being the default.

### Demo
#### Empty, default

<img width="616" alt="image" src="https://github.com/user-attachments/assets/f8955c1e-6c42-480b-9b61-0795eed5c333" />

#### Empty, PAT

<img width="609" alt="image" src="https://github.com/user-attachments/assets/c2013a20-588a-419d-8915-0ac54a7913b4" />

AlexVialaBellander committed:
Add M2M Oauth to Databricks Driver

luizarakaki commented:
Thanks @AlexVialaBellander for the PR.

We will review it.

I can see two things that we should do to merge it:

1. Add tests. Most drivers have a `can-connect?` test.
2. In the UI, we shouldn't show both options. We should show personal token by default and the user can change to use M2M OAuth. This PR has a similar change using the visible-if and a toggle https://github.com/metabase/metabase/pull/51249

AlexVialaBellander committed:
Add toggle for PAT/M2M Oauth

Introduces a 'Use Personal Access Token' toggle button which shows different UI components when setting up the connection. The implementation was inspired by @snoe in #51249

AlexVialaBellander commented:
> Thanks @AlexVialaBellander for the PR.
> 
> We will review it.
> 
> I can see two things that we should do to merge it:
> 
> 1. Add tests. Most drivers have a `can-connect?` test.
> 2. In the UI, we shouldn't show both options. We should show personal token by default and the user can change to use M2M OAuth. This PR has a similar change using the visible-if and a toggle [Set key auth as the default auth method, put warning password #51249](https://github.com/metabase/metabase/pull/51249)

Hi @luizarakaki, I've added the toggle. Not sure how the tests are supposed to be implemented (admittedly haven't looked either)

lbrdnk reviewed (commented):
Thank you for your contribution! Please see my comments and address those. The one on `use-m2m` is most important. I'll take care of the test.

AlexVialaBellander commented on modules/drivers/databricks/resources/metabase-plugin.yaml:
```diff
@@ -12,27 +12,58 @@ driver:
         helper-text: Found in SQL Warehouses > Connection details
         required: true
+        placeholder: /sql/1.0/endpoints/abcdef1234567890
+      - name: use-token
```
Good input. Would it be possible to retain existing connections while making m2m default?

AlexVialaBellander committed:
Address feedback from @ibrdnk

Addresses the feedback related to #pullrequestreview-2515818008

AlexVialaBellander commented on modules/drivers/databricks/src/metabase/driver/databricks.clj:
```diff
@@ -201,29 +201,35 @@
+                              (when (and (string? additional-options) (not (str/blank? additional-options)))
+                                (if (str/starts-with? additional-options ";")
+                                  additional-options
+                                  (str ";" additional-options))))
```
Added back.

AlexVialaBellander commented on modules/drivers/databricks/src/metabase/driver/databricks.clj:
```diff
@@ -201,29 +201,35 @@
-  (merge
-   {:classname        "com.databricks.client.jdbc.Driver"
-    :subprotocol      "databricks"
-    ;; Reading through the changelog revealed `EnableArrow=0` solves multiple problems. Including the exception logged
-    ;; during first `can-connect?` call. Ref:
-    ;; https://databricks-bi-artifacts.s3.us-east-2.amazonaws.com/simbaspark-drivers/jdbc/2.6.40/docs/release-notes.txt
```
Added back.

AlexVialaBellander commented on modules/drivers/databricks/src/metabase/driver/databricks.clj:
```diff
@@ -201,29 +201,35 @@
-    :pwd            token
-    :UserAgentEntry (format "Metabase/%s" (:tag config/mb-version-info))
-    :UseNativeQuery 1}
-   ;; Following is used just for tests. See the [[metabase.driver.sql-jdbc.connection-test/perturb-db-details]]
-   ;; and test that is using the function.
```
Added back.

lbrdnk commented on modules/drivers/databricks/resources/metabase-plugin.yaml:
```diff
@@ -12,27 +12,58 @@ driver:
+      - name: use-m2m
+        display-name: Use Machine to Machine (M2M) authentication
+        type: boolean
+        default: false
```
I believe making m2m new default is right decision. Wouldn't changing this  default put us in that position?

lbrdnk replied:
Please help me find the error in my reasoning then:
- Setting the default to true here (line 27) will make m2m auth default in a sense that screen for adding database connection will provide this option as a default. User would have to explicitly opt out for PAT auth.
- _Database details_ of connections created before this change will be transformed into jdbc specs using PAT code path because `use-m2m` is `nil` in those cases.
- New _database connections_ will have `use-m2m` set either false or true, hence the right spec would be generated.
- Old database connections will get the `use-m2m` populated on successful edit.

AlexVialaBellander replied:
In your reasoning you are assuming that existing configurations will leave use-m2m as nil/null/none with my proposed changes. If this is true, then your reasoning holds. I was operating under the assumption that it would be set to the default value. I will change the `use-m2m` var default to `true`. Please resolve this if this makes sense.

lbrdnk commented on modules/drivers/databricks/src/metabase/driver/databricks.clj:
```diff
@@ -201,29 +201,38 @@
-   ;; and test that is using the function.
-   (when log-level
-     {:LogLevel log-level})))
+  (let [base-spec
+        {
+          :classname      "com.databricks.client.jdbc.Driver"
+          :subprotocol    "databricks"
+          ;; Reading through the changelog revealed `EnableArrow=0` solves multiple problems. Including the exception logged
+          ;; during first `can-connect?` call. Ref:
+          ;; https://databricks-bi-artifacts.s3.us-east-2.amazonaws.com/simbaspark-drivers/jdbc/2.6.40/docs/release-notes.txt
+          :subname        (str "//" host ":443/;EnableArrow=0"
+                              ";ConnCatalog=" (codec/url-encode catalog)
+                              (preprocess-additional-options additional-options))
+          :transportMode  "http"
+          :ssl            1
+          :HttpPath       http-path
+          :UserAgentEntry (format "Metabase/%s" (:tag config/mb-version-info))
+          :UseNativeQuery 1}
+          ;; Following is used just for tests. See the [[metabase.driver.sql-jdbc.connection-test/perturb-db-details]]
+          ;; and test that is using the function.
+        base-spec (if log-level (assoc base-spec :LogLevel log-level) base-spec)]
+    (if use-m2m
+      ;; M2M OAuth
+      (assoc base-spec
+             :AuthMech 11
+             :Auth_Flow 1
+             :OAuth2ClientId client-id
+             :OAuth2Secret oauth-secret)
+      ;; PAT authentication
+      (assoc base-spec
+             :AuthMech 3
+             :uid "token"
+             :pwd token))))
```
nit: (1) whitespace, (2) I think rewriting this function in a form as follows would make it cleaner.
```
(merge 
  {base}
  (when log-level
    {:LogLevel log-level})
  (if use-m2m
    {:AuthMech 11,,,}
    {:AuthMech 3,,,}))
```

AlexVialaBellander replied:
I've tried to implement your feedback. Please take a look and resolve if OK

AlexVialaBellander committed:
Default M2M, Add conditional Merge, Reduce whitespace

jeff-bruemmer commented on docs/databases/connections/databricks.md:
```diff
@@ -14,22 +14,29 @@ The display name for the database in the Metabase interface.
 See [Personal Access Token (PAT)](https://docs.databricks.com/en/dev-tools/auth/pat.html).
 
-## Catalog
+### Authenticate access with a service principal using OAuth (OAuth M2M)
+See [Authenticate access with a service principal using OAuth](https://docs.databricks.com/en/dev-tools/auth/oauth-m2m.html).
```
```suggestion
### Authenticate access with a service principal using OAuth (OAuth M2M)

See [Authenticate access with a service principal using OAuth](https://docs.databricks.com/en/dev-tools/auth/oauth-m2m.html).
```

jeff-bruemmer commented on docs/databases/connections/databricks.md:
```diff
@@ -14,22 +14,29 @@ The display name for the database in the Metabase interface.
-See [Compute settings for the Databricks JDBC Driver](https://docs.databricks.com/en/integrations/jdbc/compute.html).
+Additionally, see [Compute settings for the Databricks JDBC Driver](https://docs.databricks.com/en/integrations/jdbc/compute.html).
+
+## Authentication
+There are two ways to authenticate with Databricks on Metabase. You can use a personal access token (PAT) or a service principal using OAuth (OAuth M2M).
```
```suggestion
## Authentication

There are two ways to authenticate with Databricks. You can use a personal access token (PAT) or a service principal using OAuth (OAuth M2M).
```

jeff-bruemmer commented on docs/databases/connections/databricks.md:
```diff
@@ -14,22 +14,29 @@ The display name for the database in the Metabase interface.
 ## HTTP path
 
-This is the Databrick's compute resources HTTP Path value.
+This is the Databrick's compute resources HTTP Path value. Often a SQL warehouse endpoint in the format `/sql/1.0/endpoints/abcdef1234567890`.
+Read [Connect to a SQL warehouse](https://docs.databricks.com/en/compute/sql-warehouse/index.html) article from Databricks for more information on how to create and connect to a SQL Warehouse.
```
```suggestion
This is the Databrick's compute resources HTTP Path value. This value is often a SQL warehouse endpoint in the format `/sql/1.0/endpoints/abcdef1234567890`. See [Connect to a SQL warehouse](https://docs.databricks.com/en/compute/sql-warehouse/index.html).
```

jeff-bruemmer reviewed (commented):
Thanks, @AlexVialaBellander!

lbrdnk reviewed (approved):
@AlexVialaBellander, the change looks good! Please address the docs remarks before we proceed to merge.

AlexVialaBellander committed:
Update docs/databases/connections/databricks.md

Co-authored-by: Jeff Bruemmer <jeffbruemmer@gmail.com>

AlexVialaBellander committed:
Update docs/databases/connections/databricks.md

Co-authored-by: Jeff Bruemmer <jeffbruemmer@gmail.com>

AlexVialaBellander committed:
Update docs/databases/connections/databricks.md

Co-authored-by: Jeff Bruemmer <jeffbruemmer@gmail.com>

AlexVialaBellander commented:
> @AlexVialaBellander, the change looks good! Please address the docs remarks before we proceed to merge.

Awesome 🚀 Many thanks for the help, I believe I have added the suggestions ❤️
